### PR TITLE
PMM-12894 Rolling strategy for connections utilization

### PR DESCRIPTION
**Pull request by artemgavrilov**
PMM-12894

This PR adds two features to PG exporter:

1. Exporter always uses limited amount of open connections, but within this limit it collects metrics in goroutines
2. Collectors created per-requests and use request context for underlying operations

### CODE REVIEW

**Review by BupycHuk — COMMENTED** [review]
Looks good to me, the only issue I see is about `collect[]` query parameters values.

File: cmd/postgres_exporter/datasource.go [hunk 1]
```diff
@@ -99,12 +88,42 @@ func (e *Exporter) discoverDatabaseDSNs() []string {
...
+}
 
+func (e *Exporter) scrapeDSN(ch chan<- prometheus.Metric, dsn string) error {
+	server, err := e.GetServer(dsn)
 	if err != nil {
 		return &ErrorConnectToServer{fmt.Sprintf("Error opening connection to database (%s): %s", loggableDSN(dsn), err.Error())}
 	}
+	defer server.Close()
+
+	level.Debug(logger).Log("msg", "scrapeDSN:"+dsn)
```
[the remark this answers is not in the document]
Comment (artemgavrilov) [reply cmd/postgres_exporter/datasource.go]: Removed

File: cmd/postgres_exporter/percona_exporter.go [hunk 2]
```diff
@@ -30,59 +37,184 @@ var (
...
+	EnableHRCollector       bool
+	EnableDefaultCollector  bool
+	EnableGoCollector       bool
+	EnableVersionCollector  bool
+	EnableProcessCollector  bool
+	EnablePostgresCollector bool
+}
+
+// makeRegistry creates a new prometheus registry with default and percona exporters.
+func makeRegistry(ctx context.Context, dsns []string, connSema *semaphore.Weighted, globlalCollectors map[string]prometheus.Collector, filters Filters) *prometheus.Registry {
```
[the remark this answers is not in the document]
Comment (artemgavrilov) [reply cmd/postgres_exporter/percona_exporter.go]: Fixed

File: cmd/postgres_exporter/percona_exporter.go [hunk 3]
```diff
@@ -30,59 +37,184 @@ var (
...
+				case "standard.version":
+					f.EnableVersionCollector = true
+				case "standard.default":
+					f.EnableDefaultCollector = true
+				case "standard.hr":
+					f.EnableHRCollector = true
+				case "standard.mr":
+					f.EnableMRCollector = true
+				case "standard.lr":
+					f.EnableLRCollector = true
```
[the remark this answers is not in the document]
Comment (artemgavrilov) [reply cmd/postgres_exporter/percona_exporter.go]: Wow, that was a mistake. Fixed.

**Review by BupycHuk — APPROVED** [review]

**Pull request merged**